### ISSUE TITLE
Better result for PixelArt scaling

### ISSUE THREAD

**Issue by RyanBram** [nagadomi/waifu2x#183]
Hi, Developers,

For common image, currently Waifu2x is still the best compared to other algorithm. But, if we want to scale a pixel art picture, Waifu2x tends to resulting jagged picture for the result. The following is the example:
![samurai original](https://cloud.githubusercontent.com/assets/4508389/25113123/baafe7e0-241f-11e7-8843-c8699343e9f2.png)

The picture above is the original version. When I scaled it 3 times, the result is like this:
![samurai waifu2x vs xbr](https://cloud.githubusercontent.com/assets/4508389/25113136/d9e20486-241f-11e7-9b74-308bdb4102c2.png)

The left picture is Waifu2x and the right picture is xBR. We can see that Waifu2x produce sharper result, but unfortunately it cannot detect the edge well, so we see jagged lines at the edge of the picture. In this case xBR show better result for edge detection. 

I hope there will be solution for this.

Thanks.

**Comment by nagadomi** [comment]
waifu2x is a machine learning based upscaler. It requires input and ideal output pairs for training. In the case of pixel art, it is difficult to make it. I can make it from pixel art upscalers like xBR or https://johanneskopf.de/publications/pixelart/ , but it would be better to use them directly.

**Comment by RyanBram** [comment]
I can use pixel art scaling algorithm like xBR directly, but from the results above, I learnt that Waifu2x still produce better result as it seems in the picture's line on pants and line on white shirt. The only thing that it lacks is edge detection to avoid jagged lines. I try to combine the result manually and the result seems good enough:
![waifu and xbr](https://cloud.githubusercontent.com/assets/4508389/25118068/de2b429a-243e-11e7-8be4-1e4030a79303.png)

**Comment by luigio** [comment]
While not tailored for pixel art, I find that it still produces pretty good results. In this particular case, it seems your issue is related to lack of support for alpha channel. If you fill the background with white, it detects outer edges correctly. Here's the (amazing) result that I got that way:
![pixelart waifu 1 6-1 6-2](https://cloud.githubusercontent.com/assets/28485623/25772596/4e8c028e-326f-11e7-8148-9c2a9b054900.png)

**Comment by RyanBram** [comment]
> Here's the (amazing) result that I got that way:

Wow, it just so amazing. It seems your result is what I am looking for.
I tried your method using http://waifu2x.udp.jp/, but the result still similar with mine above nor as good as you.

How you got that result?

**Comment by nagadomi** [comment]
It seems like using denoising. It will be more smooth edges but it is a side effect.

**Comment by luigio** [comment]
Yes, I selected artwork style and highest noise reduction every time.

It's not perfect, as you can see some details are lost (like the leftmost vertical line in each leg), but its intuition in the details that it kept was very good. 

I'm sure it could produce even better results if the network were trained specifically for pixel art with proper input-output pairs.

My suggestion would be to make the output picture in each pair by taking a high-resolution manga picture, reducing it to 16 colors without dithering and shrinking the resulting picture by a factor of 8 or more without color constraints. For the input picture, just reduce the input picture to the same set of 16 colors as before, without dithering. I wished I could do it myself, but my programming skills are very limited. :(

**Comment by nagadomi** [comment]
I think it is very difficult to make pixel art from illustrations or photo.
The generated input image must have the same rules as actual pixel art. That is not possible with a generic image resizing algorithm.

**Comment by luigio** [comment]
It's not quite ideal, but I'm pretty sure it's close enough. At the very least, it will always detect rigged edges and make them smooth, which is the main point.

Also, the rules true for pixel art have to be very similar to the ones I mentioned, or else our eyes wouldn't be able to interpret such images correctly.

**Comment by nagadomi** [comment]
hmm, here is my current opinion.

art/photo have antialiasing and color gradient on its own, so it is difficult to reduce color properly.
For two-color manga(no screen tone), good results may be obtained, but that data can not be used to convert colored pixel art.
One possibility is to convert from 3DCG to pixel art. In the 3D model with simple materials and no antialiasing, we do not have to consider color reduction algorithm and we can draw proper edges.
For example, mozuya's work: https://www.youtube.com/watch?v=XU_3ZxkDO7U
However, it is a hard work to generate various types of training data in that way.
Another possibility is to choose training data from danbooru's "oekaki" tag. These  images are not antialiased, It might be little easier to convert to pixel art, I think.

well, anyway, that is a troublesome task. if someone release a library or dataset, I will use it.

**Comment by luigio** [comment]
Here's an example of what I mean. I'll use this picture as a starting point:

![clarke](https://cloud.githubusercontent.com/assets/28485623/25813243/18563a94-3419-11e7-967c-e7cb8d969b60.png)

First I reduce color without dithering. I think anything over 8 colors would usually be too much, so let's say I use just five basic colors this time: red, blue, green, white and black. Here is the result:

![clarke2](https://cloud.githubusercontent.com/assets/28485623/25813294/4b96cd56-3419-11e7-9f52-50573f574c40.png)

Next I convert the picture back to RGB (i.e. no color limit) and shrink it by a factor of 8:

![clarke3](https://cloud.githubusercontent.com/assets/28485623/25813379/99efa662-3419-11e7-81b3-31fd53aec3e1.png)

This will be used as the output picture in a training pair. Note that it's perfectly anti-aliased.

For the input picture, I reduce color again without dithering to get an aliased picture. I use the same set of five colors as before. This is what I get:

![clarke4](https://cloud.githubusercontent.com/assets/28485623/25813676/b02fd8d8-341a-11e7-8ab5-1a5c75ee454f.png)

So there you have it. With enough training, the neural network will learn how to turn an aliased picture (pixel art) into an anti-aliased one, which then can be enlarged with your waifu2x algorithm.

**Comment by luigio** [comment]
This is almost exactly what I meant: https://imgur.com/gallery/YDbBZ

**Comment by bloc97** [comment]
This is an artifact of the pre-upscaling filter used in Waifu2x, if I'm not mistaken. Using a transposed convolution instead of a fixed bilinear/bicubic filter could make extremely-low resolution cases easier for the network.

**Comment by bloc97** [comment]
<img width="927" alt="x2 demonstration" src="https://user-images.githubusercontent.com/567732/32407643-64eb6b32-c162-11e7-91f8-91f3fe844da9.png">
In this example you can see that the effective "blur" for the 4x4 pixel image upscaled to 8x8 occupies a large portion of the resulting image, thus making the NN believe that it is part of the image, while the 64x64 upscaled to 128x128 is much more easily resolvable.

**Comment by bloc97** [comment]
As a workaround, it is possible to pass the original image through a xBRZ filter, downscaling it using bicubic/bilinear, and then applying Waifu2x. This is equivalent of doing a pixel-art pre-smoothing step to make the image less jagged.
This is what I obtained. Very promising!
![downscaled_cnn_sr2](https://user-images.githubusercontent.com/567732/32407797-62fbf57e-c164-11e7-8c89-ec561c5e2c4a.png)

**Comment by luigio** [comment]
I'm kind of disappointed that nobody has trained a neural net for pixel art upscaling yet. No reason why it shouldn't work, and it would be very useful for indie adventure game developers.

**Comment by RyanBram** [comment]
I think the problem is because nobody provide proper training materials yet.

I don't know how useful it will be, but maybe something like this DevinatArt gallery can be used for starting point, because it contains low res and hi res pixel art to be compared:
https://omegachaino.deviantart.com/art/Street-of-Rage-2-Axel-Pixel-upgrade-634629192

Or maybe we can use sprite from old version of King of Fighters game and compare it with latest version of King of Fighters game. Most sprites and fighting styles don't change much, the only different just the resolution.

**Comment by luigio** [comment]
> I think the problem is because nobody provide proper training materials yet.

As I mentioned before, input-output pairs for training are very easy to create from any image database.

> I don't know how useful it will be, but maybe something like this DevinatArt gallery can be used for starting point, because it contains low res and hi res pixel art to be compared:
> https://omegachaino.deviantart.com/art/Street-of-Rage-2-Axel-Pixel-upgrade-634629192

It will be very hard to get enough such pairs, and it's not what I'm looking for anyway. I like the lack of detail of pixel art, I just don't like seeing the pixels. So I'd love to have a tool for generating bigger, anti-aliased versions of pixel art images without adding extra detail. Something like xBR but more accurate.

**Comment by bloc97** [comment]
The problem is not that training such a network is hard. There's new network structures that are extremely simple, fast to train and are better than what Waifu2x was originally based upon. You could basically implement one yourself in a few dozen lines in Torch or Caffe and train it in a few hours on a modern GPU. But making an portable version that does not require Linux or the deep learning libraries is somewhat complicated.

**Comment by bloc97** [comment]
Until there's a good amount of demand for what you need, or if you are willing to give your own time and develop a NN-based pixel art upscaler, I am afraid that we will not see one in the near future.

**Comment by nagadomi** [comment]
mitaki28 has released a pixel art upscaler.
https://github.com/mitaki28/pixel-art-upscaler
It seems to be using Nearest Neighbor to create training data.

**Comment by luigio** [comment]
Thanks, looking good! (Although the demo is quite buggy.)

**Comment by luigio** [comment]
This is the result I got from it:

![warrior up](https://user-images.githubusercontent.com/28485623/37873679-ed42365e-3021-11e8-8197-c7d86a91b3af.png)

**Comment by DonaldTsang** [comment]
Here is an idea: make an ML-based pixel framer and downscaler, and then feed the results into Waifu2x, and then reverse the procedures using the same pixel grid to upscale

**Comment by Azq2** [comment]
Yet another way:
- xBRz 2x upscale -> waifu2x-chainer -> lanczos 2x downscale

2x
![image](https://github.com/nagadomi/waifu2x/assets/5139482/a1380ab1-f0a7-4807-88ce-0676d4ff6bcc)

3x
![image](https://github.com/nagadomi/waifu2x/assets/5139482/3a5b49b5-c086-4e1d-a8e4-35e56a2476c6)

Alpha channel is fine.

Command for this example:
```
./waifu2x.py -i "$input" -s 2 -T 8 -t -a 3 -n 0 -m noise_scale -c rgb -o "$output"
```

P.S. Maybe training VGG7/ResNet10 with xBRz instead of bicubic/lanczos can produce good results for pixelart?

Patch for [waifu2x chainer](https://github.com/tsurumeso/waifu2x-chainer):
```patch
--- a/waifu2x.py
+++ b/waifu2x.py
@@ -1,3 +1,4 @@
+#!/usr/bin/env python3
 from __future__ import division
 import argparse
 import os
@@ -5,8 +6,9 @@ import time
 
 import chainer
 import numpy as np
-from PIL import Image
+from PIL import Image, ImageOps, ImageFilter
 import six
+import xbrz
 
 from lib import iproc
 from lib import reconstruct
@@ -32,6 +34,9 @@ def denoise_image(cfg, src, model):
 
 
 def upscale_image(cfg, src, scale_model, alpha_model=None):
+    src = xbrz.scale_pillow(src.convert('RGBA'), 2).convert('RGBA')
+    #src = src.resize((int(src.size[0] / 2), int(src.size[1] / 2)), Image.LANCZOS)
+
     dst, alpha = split_alpha(src, scale_model)
     for i in range(int(np.ceil(np.log2(cfg.scale_ratio)))):
         six.print_('2.0x scaling...', end=' ', flush=True)
@@ -51,6 +56,7 @@ def upscale_image(cfg, src, scale_model, alpha_model=None):
             alpha = reconstruct.image(
                 alpha, alpha_model, cfg.block_size, cfg.batch_size)
         six.print_('OK')
+
     dst_w = int(np.round(src.size[0] * cfg.scale_ratio))
     dst_h = int(np.round(src.size[1] * cfg.scale_ratio))
     if dst_w != dst.size[0] or dst_h != dst.size[1]:
@@ -61,6 +67,7 @@ def upscale_image(cfg, src, scale_model, alpha_model=None):
         if alpha.size[0] != dst_w or alpha.size[1] != dst_h:
             alpha = alpha.resize((dst_w, dst_h), Image.LANCZOS)
         dst.putalpha(alpha)
+    dst = dst.resize((int(dst_w / 2), int(dst_h / 2)), Image.LANCZOS)
     return dst
 
 
@@ -94,7 +101,7 @@ def load_models(cfg):
         if os.path.exists(model_path):
             models['noise_scale'] = srcnn.archs[cfg.arch](ch)
             chainer.serializers.load_npz(model_path, models['noise_scale'])
-            alpha_model_name = 'anime_style_scale_{}.npz'.format(cfg.color)
+            alpha_model_name = 'anime_style_noise3_scale_{}.npz'.format(cfg.color)
             alpha_model_path = os.path.join(model_dir, alpha_model_name)
             models['alpha'] = srcnn.archs[cfg.arch](ch)
             chainer.serializers.load_npz(alpha_model_path, models['alpha'])
```

**Comment by nagadomi** [comment]
https://github.com/WuZongWei6/Pixelization can convert images into natural looking pixel art.
I think this model has the ability to generate better paired data than before.
However, it is non-commercial license.
